README.md
---
title: "STAT585_LAB3"
author: "Team 4"
output:
  html_document:
    keep_md: True
---

## Link
[https://github.com/lydiaPenglish/smoothMap](https://github.com/lydiaPenglish/smoothMap)

## Team Members
- Lydia English
- Yang Qiao
- Xiyuan Sun



## Description
The goal of `smoothMap` is to generate a function to help map files.

## Installation
You can install the released version of `smoothMap` from [Github](https://github.com/lydiaPenglish/smoothMap) with:
``` r
devtools::install_github("lydiaPenglish/smoothMap")
```

## Example
This is a basic example which shows you how to use the functions in `smoothMap`:


```r
library(smoothMap)
library(ggplot2)

isl_file <- system.file("extdata", "gadm36_ISL_1.shp", package = "smoothMap")
isl_data <- team_1(isl_file, tolerance = 0.01)
ggplot(isl_data, aes(x = long, y = lat, group = group)) +
  geom_polygon(color = "black", fill = "white", size = 0.2) +
  labs(x = "Longitude", y = "Latitude", title = "Iceland") +
  coord_quickmap() +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
```

<img src="README_files/figure-html/example-1.png" style="display: block; margin: auto;" />

```r
pan_file <- system.file("extdata", "gadm36_PAN_1.shp", package = "smoothMap")
pan_data <- team_4(pan_file, tolerance = 0.01)
ggplot(pan_data, aes(x = long, y = lat, group = group)) +
  geom_polygon(color = "black", fill = "white", size = 0.2) +
  labs(x = "Longitude", y = "Latitude", title = "Panama") +
  coord_quickmap() +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
```

<img src="README_files/figure-html/example-2.png" style="display: block; margin: auto;" />


NEWS.md
# smoothMap 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(smoothMap)

test_check("smoothMap")


tests/testthat/test-team_1.R
context("test-team_1")

test_that("Function team_1 doesn't work.", {
  filename <- system.file("extdata", "gadm36_ISL_1.shp", package = "smoothMap")

  expect_error(team_1("gadm36_ISL_1.shx"))
  expect_error(team_1(filename, tolerance = "tol"))
  expect_error(team_1(filename, tolerance = -1))
  expect_error(team_1(filename, tolerance = Inf))

  dat <- team_1(filename, tolerance = 0.01)
  dat_head <- structure(list(GID_0 = "ISL", NAME_0 = "Iceland", GID_1 = "ISL.1_1",
                             NAME_1 = "Austurland", VARNAME_1 = "Eastland", NL_NAME_1 = NA_character_,
                             TYPE_1 = "Landsvæði", ENGTYPE_1 = "Region", CC_1 = NA_character_,
                             HASC_1 = "IS.AL", group = 1L, long = -12.05416679, lat = 63.09583282,
                             order = 1L), row.names = c(NA, -1L), class = c("tbl_df",
                                                                            "tbl", "data.frame"))
  expect_equivalent(head(dat, 1), dat_head)
})


R/team_1.R
if(getRversion() >= "2.15.1")  utils::globalVariables(c(".", "geometry"))

#' Create a geometry data frame.
#' @description Read a shape file and create a data frame by reforming geometry column to longitude, latitude, order, group.
#'
#' See \href{https://github.com/lydiaPenglish/smoothMap}{smoothMap}.
#'
#' @param file The name of the shape (.shp) file which the data are to be read from.
#' @param tolerance The tolerance value in the metric of the input object.
#' @return A data frame.
#'
#' @importFrom methods as
#' @export
#' @author Yang Qiao
#'
#' @examples
#' filename <- system.file("extdata", "gadm36_ISL_1.shp", package = "smoothMap")
#' dat <- team_1(filename, tolerance = 0.01)
#' if (require(ggplot2))
#'   ggplot(dat, aes(x = long, y = lat, group = group)) +
#'     geom_polygon(color = "black", fill = "white", size = 0.2) +
#'     labs(x = "Longitude", y = "Latitude", title = "Iceland") +
#'     coord_quickmap()
team_1 <- function(file, tolerance = 0.1) {
  poly2df <- function(feature) {
    feature %>%
      purrr::map_depth(2, function(y) dplyr::mutate(dplyr::rename_all(data.frame(y), ~ c("long", "lat")),
                                                    order = dplyr::row_number())) %>%
      purrr::flatten() %>%
      dplyr::tibble(poly = .)
  }

  checkmate::assert_file_exists(file, extension = "shp")
  checkmate::assert_number(tolerance, lower = 0, finite = T)

  data <- sf::read_sf(file)
  oz <- maptools::thinnedSpatialPoly(as(data, "Spatial"),
                                     tol = tolerance, min = 0.001, topologyPreserve = T) %>%
    sf::st_as_sf()

  dplyr::as_tibble(oz) %>%
    dplyr::mutate(new = purrr::map(geometry, poly2df)) %>%
    dplyr::select(-geometry) %>%
    tidyr::unnest() %>%
    dplyr::mutate(group = dplyr::row_number()) %>%
    tidyr::unnest()
}


R/team_4.R
#' Generating dataframes from shapefiles for smooth maps
#' @param file A path to a shape file
#' @param tolerance A numeric value specifying the default value at which to thin the polygon. Generally less than 0.1
#' @return A dataframe with lat and long coordinates to plot a map
#'
#' @importFrom methods as
#' @export
#' @examples
#' filename <- system.file("extdata", "gadm36_PAN_1.shp", package = "smoothMap")
#' dat <- team_4(filename, tolerance = 0.01)
#' if (require(ggplot2))
#'   ggplot(dat, aes(x = long, y = lat, group = group)) +
#'     geom_polygon(color = "black", fill = "white", size = 0.2) +
#'     labs(x = "Longitude", y = "Latitude", title = "Panama") +
#'     coord_fixed()
team_4 <- function(file, tolerance){
    # initial checks
    checkmate::assertNumber(tolerance, lower = 0)
    if(tolerance > 0.5){
      warning("Tolerance above 0.5 may result in map losing features")
    }
    if(tools::file_ext(file) != "shp"){
      stop("File must be shape file (.shp) to proceed")
    }
    # function to unlist geometry tibble within shapefile
    poly2df <- function(feature) {
      feature <- unlist(feature, recursive = FALSE)
      lapply(feature, function(x) {
        dplyr::mutate(
          dplyr::rename_all(data.frame(x), ~ c("long", "lat")), order = dplyr::row_number())
      }) %>% tibble::tibble(polygon = .)
    }
    # Now read in shape file
      ozbig <- sf::read_sf(file)
      oz_st <- maptools::thinnedSpatialPoly(as(ozbig, "Spatial"),
                                            tolerance = tolerance, min = 0.001, topologyPreserve = TRUE)
      oz <- sf::st_as_sf(oz_st)
      tibble::as_tibble(oz) %>% dplyr::mutate(new = purrr::map(geometry, poly2df)) %>% dplyr::select(-geometry) %>%
      tidyr::unnest() %>% dplyr::mutate(group = dplyr::row_number()) %>% tidyr::unnest()
}

